### PR TITLE
Delayed Initialization for dash::Mutex

### PR DESCRIPTION
**Pull request by rkowalewski**
Our `dash::Mutex` did not support delayed initialization until now. There are  use cases which require a global lock handle which can then be reused within the application.

This pull request added this feature and contains additional refactoring, mainly the replacement of the native pointer member in `dash::Mutex` with a `std::unique_ptr`.

### CODE REVIEW

**Comment by fmoessbauer** [comment]
This PR looks good! However this also includes the design decision to create objects in an "invalid" state (ok, you check this by using assertions). I'm not a big fan of delayed initialisation, as this problem could be easily solved by using a global `std::unique_ptr` which is bound when initialisation is possible. This also solves many problems with recursive includes, as the size of the mutex struct does not have to be known at declaration time.

In the more general case it is questionable if all methods of an object have to be callable at all times. Ideally that should be the case, but almost no library supports that.

In short, merge it after @devreal has verified the DART stuff.

File: dash/include/dash/Mutex.h [hunk 1]
```diff
@@ -30,50 +31,75 @@ class Mutex {
...
-  
-private:
-  dart_lock_t   _mutex;
-}; // class Mutex
 
-} // namespace dash
+private:
+  dash::Team const* _team{nullptr};
+  std::unique_ptr<std::remove_pointer<dart_lock_t>::type, DestroyDARTLock>
+      _mutex{nullptr};
```
Comment (devreal) [inline dash/include/dash/Mutex.h]: I would suggest to introduce a constant `DART_LOCK_NULL` (= `(dart_lock_t*)NULL`) and use that instead of `nullptr` and `NULL` in the DASH and DART part as `dart_lock_t` is an opaque type (that happens to be `typedef`ed to a pointer).

**Review by devreal — CHANGES_REQUESTED** [review]
Thanks for taking care of this, just one comment inline

**Comment by devreal** [comment]
@fmoessbauer 

> I'm not a big fan of delayed initialisation, as this problem could be easily solved by using a global `std::unique_ptr` which is bound when initialisation is possible. 

I'm not sure I can follow you here. How would make sure the lock is initialized after the runtime initialization has been called?

**Comment by fmoessbauer** [comment]
@devreal: I meant it this way:

```cpp
std::unique_ptr<dash::mutex> mx;

int main(){
  dash::init();
  mx = std::make_unique<dash::mutex>();
  // use it

  // otherwise the mutex has to register on the global deallocation list
  mx.reset();
  dash::finalize();
  // mutex has to be deconstructed at this point in time
}
```

**Comment by rkowalewski** [comment]
> I'm not sure I can follow you here. How would make sure the lock is initialized after the runtime initialization has been called?

If I understood it correctly we can define a global `std::unique_ptr` pointing initially to a `nullptr` and reset it after runtime initialization with a new `Mutex`. While this is possible it is neither the optimal solution nor a better one IMHO because we can always reset the `unique_ptr`. It smells like misusing `unique_ptr` as a wrapper to postpone initialization.

However, in general I agree with @fmoessbauer that our concept of delayed initialization is kind of special.

**Comment by rkowalewski** [comment]
Well seems like we wrote a comment concurrently. OK to solve your use case I would expect that DART finalizes all globally allocated memory, teams, etc.

And then we just have to check if dart is initialized in the internal deleter of our `dash::Mutex`. Or am I missing something?

File: dash/include/dash/Mutex.h (hunk 1 above)
Comment (rkowalewski) [reply dash/include/dash/Mutex.h]: Good point. Maybe you can "overtake" this issue @devreal. I thought about this however, I was not sure how to define a `DART_LOCK_NULL` handle since we have opaque members in it.

Comment (devreal) [reply dash/include/dash/Mutex.h]: I will push something later today

**Comment by devreal** [comment]
I changed the implementation of the DART locks such that they are registered with the team and the segments are released upon destruction of the team. The `dash::Mutex` wrapper then still calls `dart_team_lock_destroy` to release the `malloc`'ed memory, which may happen at any time, e.g., at the end of `main`.

**Review by devreal — APPROVED** [review]

**Review by fmoessbauer — APPROVED** [review]

**Comment by fmoessbauer** [comment]
Is there still something to change, or can we merge?

**Comment by rkowalewski** [comment]
Ready to merge. ;)
